### PR TITLE
Methodology section

### PR DESCRIPTION
**Pull request by MichalisPanayides**
- Intro
- Backwards Induction
- Nash Equilibrium
- Learning Algorithms

### CODE REVIEW

**Comment by MichalisPanayides** [comment]
# Points for discussion:
- _**Nash equilibrium subsection**_: 
   - More info on how Lemke Howson works?
   - Should I mention support enumeration?
- _**Learning algorithms subsections**_: 
   - Should I include how fictitious play and stochastic fictitious play work? (currently just referenced)
- _**Missing example?**_:
   - I feel like there could be an example of the game here where Lemke-Howson and ARD is run
   - I did not include one because there is one in the _application_ section

File: sections/methodology/main.tex [hunk 1]
```diff
@@ -1,7 +1,203 @@
-\section{Methodology}
+\section{Methodology}\label{sec:methodology}
+
+Th problem defined in section \ref{sec:model_overview} describes a normal-form
+game between the decision makers of two queueing systems and a third player that 
+decides how to distribute individuals to the systems.
+The strategy space of the two queueing systems is defined as the possible values
+that the threshold parameter can take \(T_i \in (1, N_i)\).
+Then, the distributor has to decide on the proportion of individuals to send to
```
Comment (drvinceknight) [inline sections/methodology/main.tex:8]:
```suggestion
Then, the distributor has to decide on the proportion of individuals to send to 
```

File: sections/methodology/main.tex [hunk 2]
```diff
@@ -1,7 +1,203 @@
...
+The strategy space of the two queueing systems is defined as the possible values
+that the threshold parameter can take \(T_i \in (1, N_i)\).
+Then, the distributor has to decide on the proportion of individuals to send to 
+each queueing system \(p_A \text{ and } p_B\), where \(p_A, p_B \in [0, 1] \)
+and \(p_A + p_B = 1\).
+Figure \ref{fig:imperfect_info_game} form section \ref{sec:model_overview} 
+shows a diagrammatic representation of the game to be played and the decisions 
+to be made.
+As described in section \ref{sec:model_overview} queueing system \(A\) decides
+on a strategy, then queueing system \(B\) chooses its own threshold, unaware
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
on a strategy, at the same time, queueing system \(B\) chooses its own threshold, unaware 
```

File: sections/methodology/main.tex [hunk 3]
```diff
@@ -1,7 +1,203 @@
...
+Figure \ref{fig:imperfect_info_game} form section \ref{sec:model_overview} 
+shows a diagrammatic representation of the game to be played and the decisions 
+to be made.
+As described in section \ref{sec:model_overview} queueing system \(A\) decides
+on a strategy, then queueing system \(B\) chooses its own threshold, unaware 
+of the first queueing system's choice.
+Finally, the distributor makes its choice based on the strategies that the 
+queueing systems chose to play. 
+
+The problem can be formulated by 3 matrices; the two payoff matrices of the
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
The utilities to each player can be represented by 3 matrices; the two payoff matrices of the 
```

File: sections/methodology/main.tex [hunk 4]
```diff
@@ -1,7 +1,203 @@
...
+The problem can be formulated by 3 matrices; the two payoff matrices of the 
+normal form game and the routing matrix.
+The payoff matrices and their utilities are defined by equations 
+(\ref{eq:payoff_entry}) and (\ref{eq:payoff_matrices}) in section 
+\ref{sec:model_overview}.
+
+The routing matrix holds the values \((p_A, p_B)\) which are the proportion 
+of class 2 individuals to send to queueing system \(A\) and \(B\).
+Each pair \((p_A, p_B)\) can be calculated using equation 
+(\ref{eq:obj_distributor}).
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
(\ref{eq:obj_distributor}) and is essentially
a best response to the actions of the queueing systems.
```

File: sections/methodology/main.tex [hunk 5]
```diff
@@ -1,7 +1,203 @@
...
+\begin{align}
+    & \alpha L_A(p_A;T_A=1;T_B=1) + (1 - \alpha) B_A(p_A;T_A=1;T_B=1) = 
+    \nonumber \\
+    & \alpha L_B(p_B;T_A=1;T_B=1) + (1 - \alpha) B_B(p_B;T_A=1;T_B=1) 
+    \label{eq:obj_distributor_1_1}
+\end{align}
+
+The values of \(p_A\) and \(p_B\) that satisfy equation 
+(\ref{eq:obj_distributor_1_1}) can be found by using Brent's bisection 
+algorithm~\cite{brent1973algorithms}.
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
The values of \(p_A\) and \(p_B\) that satisfy equation 
(\ref{eq:obj_distributor_1_1}) are here found numerically using Brent's bisection
algorithm~\cite{brent1973algorithms}.
```

File: sections/methodology/main.tex [hunk 6]
```diff
@@ -1,7 +1,203 @@
...
+
+Similar to the routing matrix, the above procedure can be repeated for all 
+possible values of \(T_A\) and \(T_B\) to populate all entries of the payoff 
+matrices. 
+
+\subsection{Nash Equilibrium}\label{sec:methodology_nash_equilibrium}
+
+Having calculated the payoff matrices \(A\) and \(B\), several algorithms can 
+be used to help us find the Nash equilibrium of the game. 
+The Nash equilibrium is the point of the game were both players have no
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
be used to measure some form of emergent behaviour.
One possibility would be to compute the Nash equilibrium
which is the point of the game were both players have no 
```

File: sections/methodology/main.tex [hunk 7]
```diff
@@ -1,7 +1,203 @@
...
+incentive to deviate from their played strategies \cite{kreps1989nash}.
+In other words their chosen strategies are a best response to each other.
+
+The Lemke-Howson algorithm is used to get all Nash equilibria of the game 
+\cite{LemkeHowson}.
+Lemke-Howson uses best response polytopes to get one of the Nash equilibrium of
+the game. 
+By iterating over all possible dropped labels of the best response polytope all
+Nash equilibria can be acquired.
+All game theoretic calculations were done in Python using the Nashpy library
```
Comment (drvinceknight) [inline sections/methodology/main.tex]: This is not actually true @11michalis11, it is possible for the Lemke Howson algorithm to not be able to reach some Nash equilibria (for example in the battle of the sexes).

```suggestion
Computation of Nash equilibria can be done relatively efficiently using the Lemke Howson algorithm
\cite{LemkeHowson}.
Lemke-Howson uses best response polytopes to get one of the Nash equilibrium of
the game. 
Other algorithms exists that will compute all Nash equilibria but 
for large games the computational complexity becomes problematic.
All game theoretic calculations were done in Python using the Nashpy library 
```

File: sections/methodology/main.tex [hunk 8]
```diff
@@ -1,7 +1,203 @@
...
+In other words their chosen strategies are a best response to each other.
+
+The Lemke-Howson algorithm is used to get all Nash equilibria of the game 
+\cite{LemkeHowson}.
+Lemke-Howson uses best response polytopes to get one of the Nash equilibrium of
+the game. 
+By iterating over all possible dropped labels of the best response polytope all
+Nash equilibria can be acquired.
+All game theoretic calculations were done in Python using the Nashpy library 
+\cite{axelrodproject}.
```
Comment (drvinceknight) [inline sections/methodology/main.tex]: Wrong key?

File: sections/methodology/main.tex [hunk 9]
```diff
@@ -1,7 +1,203 @@
...
+
+\subsection{Learning Algorithms}\label{sec:methodology_learning_algorithms}
+
+In order to analyse the strategies played by the two EDs, the learning algorithm
+asymmetric replicator dynamics is used \cite{asymmetricreplicatordynamics}.
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion

Another approach to measuring emergent behaviour is to consider 
the emergence itself and not just stable end points. 
Indeed, some Nash equilibria might not arise naturally. Thus, in order to analyse the strategies played by the two EDs, the learning algorithm:
asymmetric replicator dynamics is used \cite{asymmetricreplicatordynamics}.
```

File: sections/methodology/main.tex [hunk 10]
```diff
@@ -1,7 +1,203 @@
...
+The two players are treated as two separate populations where each
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
The two players are modelled as two separate populations where each 
```

File: sections/methodology/main.tex [hunk 11]
```diff
@@ -1,7 +1,203 @@
...
+changes based on their previous interactions.
+The fitness of each strategy is defined as:
+
+\begin{equation}\label{eq:asymmetric_fitness}
+    f_x = Ay, \quad f_y = x^T B
+\end{equation}
+
+Here, \(x \in \mathbb{R}^{m \times 1} \) and
+\(y \in \mathbb{R}^{n \times 1}\) correspond to the proportion of individuals 
+that play each strategy on the two populations.
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
that play each strategy in each population.
```

File: sections/methodology/main.tex [hunk 12]
```diff
@@ -1,7 +1,203 @@
...
+Similarly, the average fitness of each strategy is defined as:
```
Comment (drvinceknight) [inline sections/methodology/main.tex]:
```suggestion
Similarly, the average fitness of each strategy is given by:
```

**Review by drvinceknight — CHANGES_REQUESTED** [review]
Looking good @11michalis11 :+1:

File: sections/methodology/main.tex (hunk 1 above)
Comment (MichalisPanayides) [reply sections/methodology/main.tex:8]: Okay ¯\\_(ツ)_/¯

File: sections/methodology/main.tex (hunk 8 above)
Comment (MichalisPanayides) [reply sections/methodology/main.tex]: Yep, that's what just what you used here: https://github.com/drvinceknight/Nashpy/blob/main/CITATION.md

Will change it 👍

File: sections/methodology/main.tex (hunk 1 above)
Comment (drvinceknight) [reply sections/methodology/main.tex:8]: lol I have no idea... 🤷‍♂️

File: sections/methodology/main.tex (hunk 8 above)
Comment (drvinceknight) [reply sections/methodology/main.tex]: Yeah that should definitely not say `axelrod` there 😆

**Pull request merged**